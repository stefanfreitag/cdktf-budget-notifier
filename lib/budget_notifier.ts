import { Construct } from "constructs";
import { BudgetsBudget } from "../.gen/providers/aws";

export interface BudgetNotifierProps {
  /**
   * Budget notifications will be sent to each of the recipients (e-mail addresses).
   */
  readonly recipients: Array<string>;

  /**
   * If specified the availability zones will be added as tag filter.
   */
  readonly availabilityZones?: Array<string>;
  /**
   * If specified the application name will be added as tag filter.
   */
  readonly application?: string;
  /**
   * If specified the cost center will be added as tag filter.
   */
  readonly costCenter?: string;

  /**
   * If specified the service will be added as tag filter.
   */
  readonly service?: string;
  /**
   * The threshold value in percent (0-100).
   */
  readonly threshold: number;
  /**
   * The cost associated with the budget threshold.
   */
  readonly limit: number;
  /**
   * The unit of measurement that is used for the budget threshold, such as dollars or GB.
   */
  readonly unit: string;
}

export class BudgetNotifier extends Construct {
  constructor(scope: Construct, name: string, props: BudgetNotifierProps) {
    super(scope, name);

    if (props.threshold <= 0) {
      throw new Error("Thresholds less than or equal to 0 are not allowed.");
    }

    const costFilters = this.createCostFilters(props);

    new BudgetsBudget(this, "Test", {
      budgetType: "COST",
      limitAmount: props.limit.toString(),
      limitUnit: props.unit,
      timeUnit: "MONTHLY",
      timePeriodStart: "2020-07-01_00:00",
      costFilters: costFilters,
      notification: [
        {
          comparisonOperator: "GREATER_THAN",
          threshold: props.threshold,
          thresholdType: "PERCENTAGE",
          notificationType: "ACTUAL",
          subscriberEmailAddresses: props.recipients,
        },
      ],
    });
  }

  private createCostFilters(props: BudgetNotifierProps) {
    const tags: Array<string> = [];
    if (props.application) {
      tags.push("user:Application$" + props.application);
    }

    if (props.costCenter) {
      tags.push("user:Cost Center$" + props.costCenter);
    }

    if (props.service) {
      tags.push("user:Service$" + props.service);
    }

    //Dirty fix that needs rework
    //https://github.com/terraform-providers/terraform-provider-aws/issues/5890
    if (tags.length>1){
      throw new Error("Thresholds less than or equal to 0 are not allowed.");
    }
    
    const costFilters: any = {};
    if (tags) {
      costFilters["TagKeyValue"] = props.application;
    }

    const availabilityZones: Array<string> = [];
    if (props.availabilityZones) {
      for (const az of props.availabilityZones) {
        availabilityZones.push(az);
      }
      costFilters["AZ"] = availabilityZones;
    }
    return costFilters;
  }
}
