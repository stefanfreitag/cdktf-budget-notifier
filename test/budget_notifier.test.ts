
import { Testing, TerraformStack } from "cdktf";
import { BudgetNotifier } from "../lib/budget_notifier";



test("minimal configuration", () => {  
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");
  new BudgetNotifier(stack, "team_1", {
    limit: 10,
    unit: "USD",
    recipients: ["stefan@stefreitag.de"],
    threshold: 10,
    application: 'helloWorld'
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});
