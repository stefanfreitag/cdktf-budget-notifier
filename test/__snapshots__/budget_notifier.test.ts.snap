// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal configuration 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_budgets_budget\\": {
      \\"test_team1_Test_947350B7\\": {
        \\"budget_type\\": \\"COST\\",
        \\"cost_filters\\": {
          \\"tag_key_value\\": \\"helloWorld\\"
        },
        \\"limit_amount\\": \\"10\\",
        \\"limit_unit\\": \\"USD\\",
        \\"time_period_start\\": \\"2020-07-01_00:00\\",
        \\"time_unit\\": \\"MONTHLY\\",
        \\"notification\\": [
          {
            \\"comparison_operator\\": \\"GREATER_THAN\\",
            \\"threshold\\": 10,
            \\"threshold_type\\": \\"PERCENTAGE\\",
            \\"notification_type\\": \\"ACTUAL\\",
            \\"subscriber_email_addresses\\": [
              \\"stefan@stefreitag.de\\"
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/team_1/Test\\",
            \\"uniqueId\\": \\"test_team1_Test_947350B7\\"
          }
        }
      }
    }
  }
}"
`;
